Keep App Session secrets secret

Joerger opened the pull request:
This PR updates App Session endpoints to keep App Session secrets secret to the user. These secrets are only needed by the Proxy Service to create connections to App Services on behalf of the clients. Giving the clients access to these secrets can lead to degraded security, especially with [Per-session MFA](https://github.com/gravitational/teleport/pull/39536) and [Hardware Key support](https://github.com/gravitational/teleport/pull/39469).

The only client consumer of App Session secrets before this PR was `tsh proxy azure`, which uses the app session private key to sign a JWT token auth header. This Auth header would pass through the proxy and get validated by the App Service. 

With this change, the client instead signs a JWT token with its own private key. The Proxy checks the header, verifies it with the client public key, and resigns it with the app session key. Finally, the App Service verifies and parses the JWT token with the app session public key.

Related https://github.com/gravitational/teleport/pull/39588

Joerger committed:
Don't reveal app session secrets to users.

Joerger committed:
Fix panic in WebSessionV2.WithoutSecrets.

Joerger committed:
Use client's key for azure jwt tokens

Proxy verifies and signs jwt tokens

Ensure backwards compatibility

Joerger commented:
@jentfoo If you get a chance can you let me know what you think about the `tsh proxy azure` changes? The change is summarized above.

Do you think there are any security implications to re-signing the JWT? An alternative solution would be for the Proxy service to pass the client's public key in an additional request header so that the App Service can verify the original client-signed JWT.

Tener commented:
The overall idea sounds good to me; I'll take a look at actual code ~tomorrow.

Joerger committed:
Remove unused GetAppSession methods.

Tener commented on lib/auth/auth_with_roles.go:
```diff
@@ -5300,13 +5299,29 @@ func (a *ServerWithRoles) GetAppSession(ctx context.Context, req types.GetAppSes
+		// App Session secrets should not be returned to the user. We only do this
+		// here for backwards compatibility with `tsh proxy azure`, which uses the
+		// app session key to sign JWT tokens with Azure claims. This check means
+		// that `tsh proxy azure` will fail for old clients when used with Per-session
+		// MFA or Hardware Key support, which is planned for release in v16.0.0.
+		identity := a.context.Identity.GetIdentity()
+		if identity.MFAVerified == "" || identity.PrivateKeyPolicy != keys.PrivateKeyPolicyWebSession {
+			return session, nil
 		}
```
Is there any practical way a user may "downgrade" their own session, removing the `MFAVerified` bit, yet gaining access to the private keys here?

Tener commented on lib/auth/auth_with_roles.go:
```diff
@@ -5400,7 +5415,8 @@ func (a *ServerWithRoles) CreateAppSession(ctx context.Context, req *proto.Creat
 	}
-	return session, nil
+
+	return session.WithoutSecrets(), nil
```
I'm somewhat surprised this doesn't break anything.

Tener commented on lib/web/app/handler.go:
```diff
@@ -427,9 +430,81 @@ func (h *Handler) getAppSessionFromCert(r *http.Request) (types.WebSession, erro
 		return nil, err
 	}
+
+	if identity.AzureIdentities != nil {
+		if err := h.tryResignAzureJWTCookie(r, ws); err != nil {
+			h.log.WithError(err).Debugf("failed to re-sign jwt cookie")
+		}
+	}
```
We should verify the request is targeting actual Azure, rather than some other app. I think that checking for azure identities is not sufficient.

Joerger replied:
Ah good catch, I meant to check `identity.RouteToApp.AzureIdentity` here, which IIUC should only be set for Azure app sessions. I can also add a check for the actual app if we think it's needed.

Tener commented on lib/auth/auth_with_roles.go:
```diff
@@ -5400,7 +5415,8 @@ func (a *ServerWithRoles) CreateAppSession(ctx context.Context, req *proto.Creat
 	}
-	return session, nil
+
+	return session.WithoutSecrets(), nil
```
I'm somewhat surprised this doesn't break anything.

Joerger replied:
Yeah, the session secrets weren't used anywhere after `CreateAppSession`.

Joerger replied:
`Snowflake` and `SamlIdP` sessions won't be affected by this change as they have their own unique endpoints. We would need to repeat this change to support MFA and Hardware Key support for those. From a glance, it looks like those don't use the secrets either so it should be simpler.

Tener commented on lib/auth/auth_with_roles.go:
```diff
@@ -5300,13 +5299,29 @@ func (a *ServerWithRoles) GetAppSession(ctx context.Context, req types.GetAppSes
+		// app session key to sign JWT tokens with Azure claims. This check means
+		// that `tsh proxy azure` will fail for old clients when used with Per-session
+		// MFA or Hardware Key support, which is planned for release in v16.0.0.
+		identity := a.context.Identity.GetIdentity()
+		if identity.MFAVerified == "" || identity.PrivateKeyPolicy != keys.PrivateKeyPolicyWebSession {
+			return session, nil
 		}
```
Is there any practical way a user may "downgrade" their own session, removing the `MFAVerified` bit, yet gaining access to the private keys here?

Joerger replied:
This shouldn't be possible, no. There is no way for a user to update their own session and replace its certs. Even if they could, the session would not be MFA verified so there would be no harm done.

Or is your concern that we _should_ let users downgrade sessions? If the user is using a new client, there would be no reason for them to access the private key. In v17.0.0 we will always return the session without secrets.

Tener replied:
This is probably a theoretical case, but the oddness here is that we grant more privileges (access to the private key) to a strictly weaker identity (without MFA verified bit). Hence I wonder if the user can force that downgrade, effectively increasing their access. I don't know how would that happen, hence my question... but I figure the answer is "no".

Joerger replied:
Gotcha. The way to think about this is that the stricter identities grant more privilege, hence they need to be guarded more strictly. Even if the user could downgrade the privilege, it would no longer need to be guarded as strictly.

Tener commented on lib/web/app/handler.go:
```diff
@@ -427,9 +430,81 @@ func (h *Handler) getAppSessionFromCert(r *http.Request) (types.WebSession, erro
+// token containing Azure claims signed by the client's private key. If
+// found, the token is resigned using the app session's private key so
+// that the App Service can validate it using the app session's public key.
+func (h *Handler) tryResignAzureJWTCookie(r *http.Request, ws types.WebSession) error {
```
It would be good to add some tests here.

Tener commented on tool/tsh/common/app_azure.go:
```diff
@@ -231,23 +232,17 @@ func (a *azureApp) startLocalALPNProxy(port string) error {
-		return trace.Wrap(err)
+	signer, ok := appCerts.PrivateKey.(crypto.Signer)
+	if !ok {
+		return trace.BadParameter("private key type %T does not implement crypto.Signer", appCerts.PrivateKey)
```
Can we add some hints to the error message - what should the user do in such case?

Joerger replied:
Realistically this should never fail, `appCerts.PrivateKey` should always cast to `crypto.Signer`.

```go
// "crypto" package
// 
// PrivateKey represents a private key using an unspecified algorithm.
//
// Although this type is an empty interface for backwards compatibility reasons,
// all private key types in the standard library implement the following interface
//
//	interface{
//	    Public() crypto.PublicKey
//	    Equal(x crypto.PrivateKey) bool
//	}
//
// as well as purpose-specific interfaces such as [Signer] and [Decrypter], which
// can be used for increased type safety within applications.
type PrivateKey any
```

We just do the check to ensure there isn't an unexpected panic.

Joerger committed:
Move azure jwt token resigning logic into the session transport's
rewriteRequest method with an improved Azure app check

Fix GetAppSession backward compatibility check

jentfoo reviewed (approved):
Seems like an improvement to me

> An alternative solution would be for the Proxy service to pass the client's public key in an additional request header so that the App Service can verify the original client-signed JWT

I don't believe this would add any additional security.  Ultimately you would still be reliant on what the proxy provides.  So I think having the signature done by the proxy is more explicit and likely safer.

Joerger committed:
Add test.

Joerger committed:
Address lint comments.

Joerger committed:
Address parseBearerToken comment.

Joerger commented on lib/web/app/handler_test.go:
```diff
@@ -80,7 +80,7 @@ func TestAuthPOST(t *testing.T) {
 	appCookieValue := "5588e2be54a2834b4f152c56bafcd789f53b15477129d2ab4044e9a3c1bf0f3b"
 
-	fakeClock := clockwork.NewFakeClockAt(time.Date(2017, 05, 10, 18, 53, 0, 0, time.UTC))
+	fakeClock := clockwork.NewFakeClockAt(time.Date(2017, 0o5, 10, 18, 53, 0, 0, time.UTC))
```
Looks like they work with `clockwork.NewFakeClock()`

Joerger committed:
apply suggestion about indentation.

Joerger committed:
Merge branch 'master' into joerger/protect-app-session-secrets

Joerger commented:
@zmb3 @r0mant can I please get an `/excludeflake *`?

r0mant commented:
/excludeflake *

Joerger committed:
Merge branch 'master' into joerger/protect-app-session-secrets
